[GPII-4229]: Refactor disk snapshots monitoring

natarajaya opened the pull request:
This PR:
* Removes alert policy based on LBM that looks for errors in K8s snapshots logs. Errors in logs does not really show that there is an issue – errors are normally followed by retries, and in case of intermittent GCP issues, policy will only create a lot of noise.
* Removes condition based on LBM that looks for snapshot creation events in K8s snapshots logs from alert policy that monitors CouchDB snapshots. We already have another more reliable LBM based on events from audit logs, so it does not make much sense to use another less reliable LBM.
* Adds another threshold condition to alert policy that monitors CouchDB snapshots. Condition will trigger in case there is less than 1 snapshot creation event per zone in 5 minutes. I was hoping to come up with something that would take into account deltas that differ between environments, but have not succeed with that. I also tried to use another LBM suggested by @amatas that is based on events from  `gce_snapshot` audit logs, but those events does not contain zone information, which makes existing LBM more reliable.
* Orphaned LBMs can't be removed at the same time with alert policies, because in this case Terraform fails to resolve the diff. I'll open a separate PR for this.

stepanstipl commented on gcp/modules/gcp-stackdriver-monitoring/alert_k8s_snapshots_couchdb.tf:
```diff
@@ -2,39 +2,47 @@ resource "google_monitoring_alert_policy" "k8s_snapshots_couchdb" {
-          alignment_period   = "600s"
+          alignment_period   = "300s"
           per_series_aligner = "ALIGN_SUM"
           group_by_fields    = []
```
I believe `cross_series_reducer` might need to be set, otherwise, the  `group_by_fields` will be ignored.

natarajaya replied:
Yeah, documentation is pretty vague about `cross_series_reducer` usage, and I don't observe any difference in condition behavior with and without it. However, if I try to configure same condition manually using UI, it completes `crossSeriesReducer` field (aggregation selector in UI) [automatically](https://app.google.stackdriver.com/metrics-explorer?project=gpii-gcp-prd&timeSelection=%7B%22timeRange%22:%221d%22%7D&xyChart=%7B%22dataSets%22:%5B%7B%22timeSeriesFilter%22:%7B%22filter%22:%22metric.type%3D%5C%22logging.googleapis.com%2Fuser%2Fcompute.disks.createSnapshot%5C%22%20resource.type%3D%5C%22gce_disk%5C%22%22,%22perSeriesAligner%22:%22ALIGN_SUM%22,%22crossSeriesReducer%22:%22REDUCE_SUM%22,%22secondaryCrossSeriesReducer%22:%22REDUCE_NONE%22,%22minAlignmentPeriod%22:%2260s%22,%22groupByFields%22:%5B%22resource.label.%5C%22zone%5C%22%22%5D,%22unitOverride%22:%221%22%7D,%22targetAxis%22:%22Y1%22,%22plotType%22:%22LINE%22%7D%5D,%22options%22:%7B%22mode%22:%22COLOR%22%7D,%22constantLines%22:%5B%5D,%22timeshiftDuration%22:%220s%22,%22y1Axis%22:%7B%22label%22:%22y1Axis%22,%22scale%22:%22LINEAR%22%7D%7D&isAutoRefresh=true). So I think it is safer to add it, thanks for noticing!

stepanstipl commented on gcp/modules/gcp-stackdriver-monitoring/alert_k8s_snapshots_couchdb.tf:
```diff
@@ -2,39 +2,47 @@ resource "google_monitoring_alert_policy" "k8s_snapshots_couchdb" {
+          per_series_aligner = "ALIGN_SUM"
+
+          group_by_fields = [
+            "metric.labels.zone",
```
I believe `cross_series_reducer` might need to be set, otherwise, the  `group_by_fields` will be ignored.

natarajaya replied:
Answered above.

stepanstipl commented on gcp/modules/gcp-stackdriver-monitoring/alert_k8s_snapshots_couchdb.tf:
```diff
@@ -2,39 +2,47 @@ resource "google_monitoring_alert_policy" "k8s_snapshots_couchdb" {
+        duration        = "300s"
+
+        aggregations {
+          alignment_period   = "300s"
```
Why is the alignment of 300s used here? Same as previous.

natarajaya replied:
Answered above.

stepanstipl commented on gcp/modules/gcp-stackdriver-monitoring/alert_k8s_snapshots_couchdb.tf:
```diff
@@ -2,39 +2,47 @@ resource "google_monitoring_alert_policy" "k8s_snapshots_couchdb" {
+        }
+
+        denominator_filter       = ""
+        denominator_aggregations = []
```
Why are these 2 values set to "", resp. []?

natarajaya replied:
I guess this is a question to @amatas, these two values are present in every threshold-based alert policy. I don't think that anything changes if I remove them, so looks like they are set to defaults and can be safely removed.

natarajaya commented on gcp/modules/gcp-stackdriver-monitoring/alert_k8s_snapshots_couchdb.tf:
```diff
@@ -2,39 +2,47 @@ resource "google_monitoring_alert_policy" "k8s_snapshots_couchdb" {
+
+        comparison      = "COMPARISON_LT"
+        threshold_value = 1.0
+        duration        = "300s"
```
The metric never goes below threshold  for both ephemeral and static environments in case `k8s-snapshots` operates normally. I didn't notice any flapping of the alert policy that deployed in my dev environment.

stepanstipl replied:
I reacted below, but marking this as resolved since it's been increased to `10m`.

natarajaya commented on gcp/modules/gcp-stackdriver-monitoring/alert_k8s_snapshots_couchdb.tf:
```diff
@@ -2,39 +2,47 @@ resource "google_monitoring_alert_policy" "k8s_snapshots_couchdb" {
     {
       condition_absent {
         filter   = "metric.type=\"logging.googleapis.com/user/compute.disks.createSnapshot\" resource.type=\"gce_disk\""
         duration = "300s"
```
Same here.

natarajaya commented on gcp/modules/gcp-stackdriver-monitoring/alert_k8s_snapshots_couchdb.tf:
```diff
@@ -2,39 +2,47 @@ resource "google_monitoring_alert_policy" "k8s_snapshots_couchdb" {
 
         aggregations {
-          alignment_period   = "600s"
+          alignment_period   = "300s"
```
Alignment window comes from backup deltas – we need to make sure that at least one snapshot per zone is created in 5 minutes. And duration setting makes alert policy more resilient to intermittent delays in snapshot processing. Are you suggesting that 10 minute window is too short or too long?

stepanstipl replied:
See below for more details, I'm suggesting that duration should be longer than only one datapoint, otherwise it'll be prone to flapping. So with alignment of `1m` I would suggest duration of `6m+` (as the original data points are still expected to arrive in 5m intervals).

natarajaya replied:
There are more than a single snapshot creation event occurs per zone in 5 minutes, you can see the metric [rarely goes below 2](https://app.google.stackdriver.com/metrics-explorer?project=gpii-gcp-prd&timeSelection=%7B%22timeRange%22:%226h%22%7D&xyChart=%7B%22dataSets%22:%5B%7B%22timeSeriesFilter%22:%7B%22filter%22:%22metric.type%3D%5C%22logging.googleapis.com%2Fuser%2Fcompute.disks.createSnapshot%5C%22%20resource.type%3D%5C%22gce_disk%5C%22%22,%22perSeriesAligner%22:%22ALIGN_SUM%22,%22crossSeriesReducer%22:%22REDUCE_SUM%22,%22secondaryCrossSeriesReducer%22:%22REDUCE_NONE%22,%22minAlignmentPeriod%22:%22300s%22,%22groupByFields%22:%5B%22resource.label.%5C%22zone%5C%22%22%5D,%22unitOverride%22:%221%22%7D,%22targetAxis%22:%22Y1%22,%22plotType%22:%22LINE%22%7D%5D,%22options%22:%7B%22mode%22:%22COLOR%22%7D,%22constantLines%22:%5B%5D,%22timeshiftDuration%22:%220s%22,%22y1Axis%22:%7B%22label%22:%22y1Axis%22,%22scale%22:%22LINEAR%22%7D%7D&isAutoRefresh=true) for prd. I don't mind giving it more time to be on the safe side, but I observe no flapping whatsoever for the alert policy that is deployed to my environment (in 1+ week).

stepanstipl replied:
> There are more than a single snapshot creation event occurs per zone in 5 minutes

Yeah, so that seems quite weird, and we probably need to improve the metric. I inspected it a bit more detail and there are several issues:

**a) 2 events seem to be created for every snapshot**, so our metric is effectively double of what we expect, check for example following filter in `prd`:
```
resource.type="gce_disk" AND protoPayload.methodName="v1.compute.disks.createSnapshot"
operation.id="operation-1575050294434-5987ff94d9b78-5f480ee8-de318034"
```

<img width="991" alt="Screenshot 2019-11-29 at 18 04 45" src="https://user-images.githubusercontent.com/3027227/69884957-c7a2f880-12d2-11ea-95c7-1f02e5c8a886.png">

It's 2 operations for one disk - one seems to be logged when the snapshot starts and other one when the snapshot is finished. There are some more differences in metadata, but interesting seems to be `operation.first: true` in first and the `operation.last: true` in the other one, perhaps that might help.

**b) the metric will currently  count errors when the snapshots fail too** - try following filter:
```
resource.type="gce_disk" AND protoPayload.methodName="v1.compute.disks.createSnapshot"
severity="ERROR"
```
it means that snapshot failed, but it will still be counted against the LBM which only has a filter `resource.type="gce_disk" AND protoPayload.methodName="v1.compute.disks.createSnapshot"`

**c) the metric will capture events from any snapshots** - although we don't have any other regular snapshots atm., if we ever do, the metric will be skewed. Perhaps it would make sense to filter further, otherwise, there's no guarantee that the snapshots even belong to CouchDB. Potentially interesting fields:
`protoPayload.authenticationInfo.principalEmail="gke-cluster-pod-k8s-snapshots@gpii-gcp-prd.iam.gserviceaccount.com"`, or `protoPayload.request.name="pv-database-storage-couchdb-couchdb-2-291119-181318"` although the later one is only present in the first (snapshot start) event.

> but I observe no flapping whatsoever

In addition to the above, have you tested situation like `k8s-snapshots` update, or GKE Node pool update when the "normal" operation of `k8s-snapshots` is disrupted for a whiile?

natarajaya replied:
@stepanstipl Thanks for the research, it is very helpful! GitHub makes it pretty difficult to keep inline conversations, I'll reply in thread instead.

natarajaya commented:
@stepanstipl Thanks for the review, I addressed some of the points and answered other inline questions.

> I think the desired effect would be better achieved with keeping the alignment_period to 60s

Not sure I understand this, can you please elaborate?

stepanstipl commented:
Thanks for the updates @natarajaya. 👍 

>> I think the desired effect would be better achieved with keeping the alignment_period to 60s

> Not sure I understand this, can you please elaborate?

Yep, I believe that the alignment of 300s only delays the potential alert.

For example, given following incoming values:
```
|1|2|3|4|5|6|7|8|9|0|1|2|3|4|5|6|7|8|9|0| < time
|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|
|1|_|_|_|_|1|_|_|_|_|_|_|_|_|_|_|_|_|_|_| < incoming data timeseries a
 [-------] [-------] [-------] [-------]  < alignment blocks 5m
|_|_|_|_|1|_|_|_|_|1|_|_|_|_|_|_|_|_|_|_| < aligned data for a (reduced cross-series will look same)
                     ^ missing data count starts here (5m)
                              ^ alert is created
```
With the alignment period of 5m the alert will be created 10m after the metric stops coming.

Now same situation with alignment of 1m:
```
|1|2|3|4|5|6|7|8|9|0|1|2|3|4|5|6|7|8|9|0| < time
|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|
|1|_|_|_|_|1|_|_|_|_|_|_|_|_|_|_|_|_|_|_| < incoming data timeseries a
 [][][][][][][][][][][][][][][][][][][][] < alignment blocks 1m
|1|_|_|_|_|1|_|_|_|_|_|_|_|_|_|_|_|_|_|_| < aligned data for a (reduced cross-series will look same)
             ^ missing data count starts here (5m)
                      ^ alert is created
```
The alert will be created 5m after the metric stops coming.

There's no need to use the higher alignment period here since we're alerting either on the absence of metric or value being below one, and there's the con of delayed alerts.

It also makes sense that the duration covers more than one alignment window because the above effect will also cause delayed recovery of the alert and will always trigger based on a single value (as not more than one value will be produced in the given window).

TL;DR: I believe good combination would be for example **alignment of 1m and duration of 6+m** -> this wouldn't cause the delay in alert rigger & recovery, and wouldn't flap in case the snapshots is slightly late.

natarajaya commented:
> TL;DR: I believe good combination would be for example alignment of 1m and duration of 6+m

@stepanstipl Thanks, that was pretty extensive explanation :)

So first – regarding flapping, as I explained in the inline thread above, I don't think current configuration is prone to flapping (not a single flapping event observed in a week in my dev env), but I don't mind to increase condition duration to stay on the safe side.

Second – regarding alert sensitivity, since this alert is not for a customer-facing service, I don't think it matters much if it is received 5 minutes sooner or 5 minutes later, no immediate action is expected from an operator anyway. 

I don't like the idea of using 1 minute alignment window, mostly because it's visual representation is much harder to understand:
<img width="1038" alt="Screen Shot 2019-11-28 at 12 12 37 PM" src="https://user-images.githubusercontent.com/917560/69830668-bd053800-11da-11ea-9ec5-96def4cd472a.png">

And a 5 minute alignment for comparison:
<img width="1039" alt="Screen Shot 2019-11-28 at 12 22 39 PM" src="https://user-images.githubusercontent.com/917560/69830719-e1f9ab00-11da-11ea-8c26-96179fb27022.png">

So I would prefer to leave alignment as it is and increase conditions duration to make alert less sensitive.

stepanstipl commented:
Thanks @natarajaya 

> mostly because it's visual representation is much harder to understand

I would probably prefer faster alerts & recovery over "nicer" visual representation since we're unlikely to ever look at it unless there's an alert triggered, but the increased duration should work too 👍.

I think the last bit before this is ready is that we need to improve the metric, as atm. we would count failed snapshots too.

natarajaya commented:
@stepanstipl Ok, I believe I am done with LBM refactoring:
* New `compute.disks.createSnapshot.couchdb` metric now tracks only snapshot start events (`operation.first: true`) for only CouchDB PVs. This seems less reliable, but unfortunately snapshot completion events do not contain any PV information.
*  But another LBM now tracks error events in snapshot audit logs, and there is an alert policy configured, so even if something goes wrong we should get notified. This makes me feel a little better about the fact that we do not track snapshot finalization events. 

> In addition to the above, have you tested situation like k8s-snapshots update, or GKE Node pool update when the "normal" operation of k8s-snapshots is disrupted for a whiile

Yes, pod rescheduling takes pretty short time for `k8s-snapshots`, I did notice any impact on the metric.

stepanstipl commented:
thanks for the improvemanets @natarajaya 

> New compute.disks.createSnapshot.couchdb metric now tracks only snapshot start events (operation.first: true) for only CouchDB PVs. This seems less reliable, but unfortunately snapshot completion events do not contain any PV information.

Great, I think this is definitely better! 👍 

> But another LBM now tracks error events in snapshot audit logs, and there is an alert policy configured, so even if something goes wrong on GCP side, we should still get notified. This makes me feel a little better about the fact that we do not track snapshot completion events.

This (with `threshold_value = 0.0` and `duration = "0s"`) is potentially super noisy, it will trigger on every single alert. Which, In-line with the aim to reduce alert noise, is undesirable. We don't care about intermittent error(s) in the process, we care if the snapshots are created in the end or not.

At least we should increase `duration`/`threshold` values to avoid alerting on a single error, but I would consider alerting on the number of finished snapshots. (There's an option to use `denominatorFilter` & `denominatorAggregations` in the conditions which could help track fraction of failed snapshots).

On implementations side of things:
- is the original LBM `compute.disks.createSnapshot` still used for something?
- instead of creating several LBMs (`compute.disks.createSnapshot.couchdb` & `compute.disks.createSnapshot.error`), have you considered creating one LBM with labels (like disk_id, severity or first/last operation for examle)?

natarajaya commented:
@stepanstipl Please review latest changes:
* I refactored existing LBM to extract PV name and severity labels, instead of using multiple LBMs. Labels appeared to be really nice feature, and they are supported by TF and Stackdriver UI natively, though I am not sure if this has any benefits over multiple LBM approach.
* I used denominator metric to alert only if ratio of errors exceeds 5% instead of alerting on fixed threshold. This feature is not supported in Stackdriver UI, so conditions that use denominator look pretty confusing there, however I tested this approach with another more dynamic metric, and it seems to work.

natarajaya commented:
@stepanstipl Added in-progress label, because PV name extraction is broken in LBM. I think it is possible to have per-PV time-series in LBM and have one generic alert policy for snapshots.

natarajaya commented:
And some more changes:
* Fixed PV name extraction in LBM.
* Combined alert policies into one with multiple conditions.
* Time series are now grouped by `pv_name`, so alert is now being fired per individual PV, not per zone! I had to add PV name filter regexp, otherwise policy keeps alerting on missing metrics, even for retired PVs that we no longer have.

I deployed new changes to my env, everything looks good so far!

stepanstipl commented on gcp/modules/gcp-stackdriver-monitoring/alert_disk_snapshots.tf:
```diff
@@ -0,0 +1,79 @@
+
+        comparison      = "COMPARISON_GT"
+        threshold_value = 0.05
+        duration        = "0s"
```
Shouldn't this duration be higher than `0s`, to prevent flapping?

I would consider setting it to `600s` to be consistent with the other alerts, that way it would cover more than 1 value and only alert if the issue is not resolved. Otherwise, this can be triggered by a single `ERROR` event.

natarajaya replied:
Policy will be triggered in case amount of errors exceed 5% of total amount of snapshots for the alignment period. But in case total amount of snapshots is low, of course the single error event is enough. I don't think increasing duration is desirable here, because in this case policy will be ignoring short error bursts, so I would increase threshold and alignment window instead if we want it to be less sensitive.

stepanstipl replied:
This is the same problem as already discussed if `duration` does not cover more than one window.

Let's take an example of a single error, and assume snapshots are suppoed to happen every 5m:
```
|1|2|3|4|5|6|7|8|9|0|1|2|3|4|5|6|7|8|9|0| < time
|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|
|1|_|_|_|_|_|_|_|_|_|_|_|_|_|_|_|_|_|_|_| < incoming data timeseries (errors)
|1|_|_|_|_|1|_|_|_|_|1|_|_|_|_|1|_|_|_|_| < incoming data timeseries (snapshots)
 [-----------------] [-----------------]  < alignment blocks 10m
|_|_|_|_|_|_|_|_|_|5|_|_|_|_|_|_|_|_|_|0| < aligned data after denominator (in 10s of %)
                    ^ alert triggers immediately
                                        ^ alert is recovered
```
^ with `0` duration and `10m` alignment, single error triggered the alert, 5m after there were successful snapshots without any further errors, the alert is delayed 5m and recovery basically 15m.


```
|1|2|3|4|5|6|7|8|9|0|1|2|3|4|5|6|7|8|9|0| < time
|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|
|1|_|_|_|_|_|_|_|_|_|_|_|_|_|_|_|_|_|_|_| < incoming data timeseries (errors)
|1|_|_|_|_|1|_|_|_|_|1|_|_|_|_|1|_|_|_|_| < incoming data timeseries (snaprshots)
 [-------] [-------] [-------] [-------]  < alignment blocks 10m
|_|_|_|_|10_|_|_|_|0|_|_|_|_|0|_|_|_|_|0| < aligned data after denominator (in 10s of %)
```
^ with `10m` duration and `5m` alignment, a single error did not trigger the alert (if there would be 2 consecutive errors, the alert would be triggered)

Now there might be another issue (and I believe there is) with how SD treats metrics in absence of data. I believe that if either of `filter` or `denominator_filter` results in no metrics, SD will not produce a data point. Now when SD evaluates alerts, it always looks at last present datapoint. This would mean that in case of error, these alerts are never recovered (imagine `empty` space instead of `0` in the above examples). So each alert would get "stuck" as we would only get data points when there's an error. I'm not sure how resolve this atm., but I'll think about it a bit more.

natarajaya replied:
There are obvious disadvantages to both approaches: with the `duration` set to include at least two data points, condition would ignore short error bursts:
```
|1|2|3|4|5|6|7|8|9|0|1|2|3|4|5|6|7|8|9|0| < time
|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|
|_|_|_|_|1|_|_|_|_|_|_|_|_|_|1|_|_|_|_|_| < incoming data timeseries (errors)
|1|_|_|_|_|1|_|_|_|_|_|_|_|_|_|1|_|_|_|_| < incoming data timeseries (snapshots)
 [-------] [-------] [-------] [-------]  < alignment blocks 5m
|_|_|_|_|100|_|_|_|_|0|_|_|_|_|100|_|_|_|_|0| < aligned data after denominator
         ^ metric is above threshold, 100% snapshots failed
         		   ^ metric recovered within duration interval, no alert triggered
```

And yes, with `duration` set to 0, condition is more sensitive, when overall number of events is low per alignment period. 

I am still slightly in favor of latter approach, but I also feel fair amount of over-engineering here – remember, policy deployed to my dev env and I observe 0 false positives for 1+ week. So I think it is fine to deploy this as-is, we can fine tune some parameters later if needed. 

> Now there might be another issue (and I believe there is) with how SD treats metrics in absence of data

Yeah, I remember we bumped into this "feature" of Stackdriver couple of times with other policies. I'll try to reproduce this to confirm, but not sure what can be done.

stepanstipl replied:
>  condition would ignore short error bursts

I think this would be desirable, if the errors stop happening I don't think we want to care

natarajaya commented:
@stepanstipl Confirming that error ratio condition is suffering from the same metric absence issue :(

stepanstipl commented:
> @stepanstipl Confirming that error ratio condition is suffering from the same metric absence issue :(

Yep, this is not great, we hit this again and again. I think we can merge this either as it is, or without the alert rule for errors altogether (as the other 2 rules already filter for `NOTICE`) and tune it later if it's not sufficient/too noisy.
